feat(aws-cloudwatch): query results widget

OliverGavin opened the pull request:
### Commit Message
feat(cloudwatch): LogGroup Query Widget

Add a `LogQueryWidget` to generate a dashboard widget
showing the results of a query from Logs Insights.
This was a missing feature which is available in
the console.

closes #3681

### End Commit Message

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 3a819db5c82db26093edd30fbb7cfaeccb2b6d3e
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=58f36142-c445-4b73-84ce-8b616b194535%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

rix0rrr commented on packages/@aws-cdk/aws-cloudwatch/lib/query.ts:
```diff
@@ -0,0 +1,75 @@
+  /**
+   * Log group to query
+   */
+  readonly logGroup: string;
```
Feels like this should accept an `ILogGroup`.

OliverGavin replied:
Ok I'll check that out, I'm pretty new with CDK

OliverGavin replied:
This would add a cyclic dependency:

```
lerna ERR! ECYCLE @aws-cdk/aws-cloudwatch -> @aws-cdk/aws-logs -> @aws-cdk/aws-cloudwatch
```

So I'm think that renaming it to `logGroupName` and keeping it a string would be the best course of action

rix0rrr replied:
I'd really really like users to be able to pass the object though. We will have to follow the existing dependency arrows while coming up with our solution. I'm thinking something like this:

```ts
// cloudwatch
interface IQueryLogGroup {
  public readonly logGroupName: string;
}

// logs
interface ILogGroup extends cloudwatch.IQueryLogGroup {
 // ...
}
```

OliverGavin replied:
Ok so I rebased and added an `IQueryLogGroup`.

Is the `ILogGroup extends cloudwatch.IQueryLogGroup` required? For TypeScript an `ILogGroup` can be supplied where an `IQueryLogGroup` is required without this inheritance (it is sort of an ugly dependency - I don't know if such a thing already exists in CDK).

Perhaps it is required for other languages using JSII etc? It might be a fair trade off to not add this inheritance (until the full solution is decided). I'm thinking that would only impact users who are using more type strict languages like Java but even so an anonymous class could still work.

The other option is to make `QueryWidgetProps#logGroup` take an `IQueryLogGroup` or a `string` and at least then with TypeScript you can pass the `ILogGroup` and in other languages you still have the option to pass the name directly as a `string`?

Let me know which you prefer and I will make the change.

rix0rrr replied:
> Perhaps it is required for other languages using JSII etc?

Exactly. This class could not be used in typed languages that use nominal subtyping (Java, C#), which would be a shame. Treating them as second class citizens (by forcing them to custom-implement an anonymous subclass) is not really acceptable in my opinion. It's not very discoverable either, and would present a big hurdle to adoption.

I agree with you that this solution is not fantastic, but it's at least pragmatic and works well enough until we have better alternatives.

> The other option is to make `QueryWidgetProps#logGroup` take an `IQueryLogGroup` or a `string`

You are thinking a union type here, which we as a rule don't use in L2 constructs because they won't translate well to Go.

I can understand you feeling dissatisfied with the available options. If it helps, I will take all of the design blame :).

OliverGavin replied:
I mean I don't personally mind; you have the bigger picture here. So will I go with what you originally proposed?

rix0rrr replied:
Yes please

OliverGavin replied:
Done and rebased

(sorry about the noise below; I'm having issues building the latest version for some reason)

rix0rrr commented on packages/@aws-cdk/aws-cloudwatch/lib/query.ts:
```diff
@@ -0,0 +1,75 @@
+        view: 'table',
+        title: this.props.title,
+        region: this.props.region || cdk.Aws.REGION,
+        query: cdk.Fn.join('', [
```
You can use a string substitution here using backticks:

```
`SOURCE '${this.logGroup.logGroupName}' | ${this.props.query}`
```

rix0rrr replied:
Does it need to be the log group name or the ARN, btw?

rix0rrr replied:
Can it be in a different account?

OliverGavin replied:
It needs to be the name. I haven't seen any way to do it cross account.

rix0rrr commented on packages/@aws-cdk/aws-cloudwatch/lib/query.ts:
```diff
@@ -0,0 +1,75 @@
+ */
+export interface QueryWidgetProps {
+  /**
+   * Title for the graph
```
Everywhere in the docs where it says "graph" it should probably say "query widget" or just "widget".

OliverGavin replied:
Right, missed that. I'll fix it in the next revision.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: f7c21e39089accc3649af8329f68b792e54cf966
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=af8bb29e-93a7-44d3-af44-bf9d6a5b0985%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c862e3cff8da4456d1ae735df05a088c974902a7
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3c2e71f8-4142-4667-9a60-d18fdeffb6e2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d33df6be37bf85f3f84b0f093e175907f14ff653
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=21addb74-8bc3-48df-9cb4-8e0325ac044a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

rix0rrr commented on packages/@aws-cdk/aws-cloudwatch/README.md:
````diff
@@ -274,6 +274,18 @@ dashboard.addWidgets(new TextWidget({
+```ts
+dashboard.addWidgets(new QueryWidget({
+  logGroup: LogGroup.fromLogGroupName(stack, 'MyLogGroup', 'my-log-group'),
+  query: `fields @message
````
One more thing, can you rename `query` to `queryString` ? I feel it will help us in the future if we ever want to provide query builder objects, which can then be used as `query: Query.fromFields(...)` or whatever.

OliverGavin replied:
Done!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 9507731f61f880452ef3b861a59ca3778be20744
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=356d2aea-5fd9-49b0-aa75-e1a52658e229%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 23e93ca1aadec5ac8e86e81ae3b990bb1e21d479
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7d05f106-3db1-424c-aefb-095cf3d69d22%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 300a79f82d44e070e32ad4dbcc4755f57f324906
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3d1979e7-efe5-4e53-aa1c-edafe2e9af6f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 867de12aa2397436bbbb1ece1539305b40a80767
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2f7f825a-bcf7-4611-8010-b00ad9492b75%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

rix0rrr commented:
You know what? You were right the first time. Just specifying the log group names makes way more sense.

I added in a convenience property for specifying the query string as well, because that whitespace sensitivity is something we can make a lot easier :).

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 2826c8c962212657a6ed4c3cec6cf972c4ef6602
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b5e9c7d2-3763-407d-b9b9-7bbdca32a50c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 17bfbe291502408ee026769c66e4af438fe091a1
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9e947618-090c-4ee3-bb0b-e655d1bd8640%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 94635cf60e8a95e293c6acc242b9c8571b3358ff
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e8f906b7-97f3-41c7-b581-1e021597f2ad%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: afc36c3f5bf36094676848ce6bdfbc8272fd2828
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4c79ae31-d53c-42dd-ba69-af70e66bde07%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

OliverGavin commented:
> You know what? You were right the first time. Just specifying the log group names makes way more sense. My apologies for making you change this.
> 
> I added in a convenience property for specifying the query string as well, because that whitespace sensitivity is something we can make a lot easier :).

Awesome, no worries. I'm using `dedent` for the whitespacing in my own project but I didn't want to add the dependency here of course. But the query lines idea is a nice trade off to have too.

OliverGavin commented:
> 
> ### AWS CodeBuild CI Report
> 
> * CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
> * Commit ID: afc36c3f5bf36094676848ce6bdfbc8272fd2828
> * Result: FAILED
> * [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4c79ae31-d53c-42dd-ba69-af70e66bde07%2Fbuild.log) (available for 30 days)
> 
> *Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*
> 
> 
> <!--
> CREATED BY GITHUB-CODEBUILD-LOGS
> -->
> 
> 

You might want to add the logGroupName I moved out of ILogGroup back in ;) (I moved it to the IQueryLogGroup)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 7835c7dcdb7e71bfa2e78ee2ec42b5af0e781813
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ce5fc6cd-e1c0-43cf-b157-9b61016337d5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 84ef063e86891c87e3d06f7a2a212792435da598
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ab64ba6d-3b1d-46ba-8c9a-703dd0b80d50%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d71d82e4a517a3aeac6304611d5b7c486b0c92eb
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=26887689-efc0-4b92-abb9-3debe8287a94%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
